devmand/gateway/src/devmand/channels/cli/KeepaliveCli.cpp

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2019-present, Facebook, Inc.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree. An additional grant
+// of patent rights can be found in the PATENTS file in the same directory.
+
+#include <devmand/channels/cli/Command.h>
+#include <devmand/channels/cli/KeepaliveCli.h>
+#include <magma_logging.h>
+
+namespace devmand::channels::cli {
+
+using devmand::channels::cli::Command;
+using namespace std;
+using namespace folly;
+
+KeepaliveCli::KeepaliveCli(
+    shared_ptr<Cli> _cli,
+    shared_ptr<Executor> _parentExecutor,
+    shared_ptr<ThreadWheelTimekeeper> _timekeeper,
+    Command&& _keepAliveCommand,
+    chrono::milliseconds _heartbeatInterval,
+    chrono::milliseconds _backoffAfterKeepaliveTimeout)
+    : cli(_cli),
+      timekeeper(_timekeeper),
+      parentExecutor(_parentExecutor),
+      serialExecutorKeepAlive(SerialExecutor::create(
+          Executor::getKeepAliveToken(_parentExecutor.get()))),
+      keepAliveCommand(_keepAliveCommand),
+      heartbeatInterval(_heartbeatInterval),
+      backoffAfterKeepaliveTimeout(_backoffAfterKeepaliveTimeout) {
+  assert(_keepAliveCommand.skipCache());
+  shutdown = false;
+
+  //  sendKeepAliveCommand();
+  MLOG(MDEBUG) << "KACli initialized";
+}
+
+KeepaliveCli::~KeepaliveCli() {
+  MLOG(MDEBUG) << "~KACli";
+  shutdown = true;
+  serialExecutorKeepAlive = nullptr;
+  parentExecutor = nullptr;
+  cli = nullptr;
+  timekeeper = nullptr;
+  MLOG(MDEBUG) << "~KACli done";
+}
+
+void KeepaliveCli::sendKeepAliveCommand() {
+  if (shutdown)
+    return;
+  MLOG(MDEBUG) << "KACli: sendKeepAliveCommand()";
+  Future<string> result = via(serialExecutorKeepAlive).thenValue([=](auto) {
```

Comment:
maybe change [=] to explicit args?

Reply:
fixed in `Guard concurrent access of kacli, qcli, rcli using shared_from_this` commit